REGRESSION(283753@main): Video as image source with `decoding="async"` is decoded always synchronously

shallawa opened the pull request:
#### f12f3a355c5a6564a9cb3845df4fbbface682933
<pre>
REGRESSION(283753@main): Video as image source with `decoding=&quot;async&quot;` is decoded always synchronously
<a href="https://bugs.webkit.org/show_bug.cgi?id=295066">https://bugs.webkit.org/show_bug.cgi?id=295066</a>
<a href="https://rdar.apple.com/153915118">rdar://153915118</a>

Reviewed by Simon Fraser.

The bug is in RenderBoxModelObject::decodingModeForImageDraw() when we choose the
DecodingMode for the image and the element has the attribute decoding=&quot;async”.
In this case we try to be smart and override the user choice if it is going to
cause flickering. We check if the element is in the viewport and if it is, we do
use DecodingMode::Synchronous.

But this should logic not be applied to animated images (or videos as images)
because for animated images we calculate the decoding mode for the next frame.

* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):

Canonical link: <a href="https://commits.webkit.org/296706@main">https://commits.webkit.org/296706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26afd37ca688ffd1f4bb4cae785d344476e1dff7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109331 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28989 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19418 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114536 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59584 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111294 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37578 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83091 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112279 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23620 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98475 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63546 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23009 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16617 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59163 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92988 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16659 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117649 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36372 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26926 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92103 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36744 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94736 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91916 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23411 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36842 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14593 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32192 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36268 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41750 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35946 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39279 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37644 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -344,7 +344,7 @@ DecodingMode RenderBoxModelObject::decodingModeForImageDraw(const Image& image,
         // will not cause flickering.
         if (imgElement->decodingMode() == DecodingMode::Asynchronous) {
-            if (bitmapImage->isAsyncDecodingEnabledForTesting())
+            if (bitmapImage->isAsyncDecodingEnabledForTesting() || bitmapImage->isAnimated())
```
It's OK to change behavior for animated GIF/APNG etc?

shallawa replied:
I think this change is restoring the code to what it was before 283753@main. If the page specifies  decoding="async" for an animated image, I think we should respect this choice since it isn't going to cause any flickering. Please notice that for animated images, the returned `DecodingMode` here is for the next frame (not the current one). 

For async animated image decoding, we always display the current frame which we ensure it is cached. The animation is advanced only when the frame finishes decoding. The image is repainted when the animation is advanced.

shallawa committed:
REGRESSION(283753@main): Video as image source with `decoding="async"` is decoded always synchronously
https://bugs.webkit.org/show_bug.cgi?id=295066
rdar://153915118

Reviewed by Simon Fraser.

The bug is in RenderBoxModelObject::decodingModeForImageDraw() when we choose the
DecodingMode for the image and the element has the attribute decoding="async”.
In this case we try to be smart and override the user choice if it is going to
cause flickering. We check if the element is in the viewport and if it is, we do
use DecodingMode::Synchronous.

But this should logic not be applied to animated images (or videos as images)
because for animated images we calculate the decoding mode for the next frame.

* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):

Canonical link: https://commits.webkit.org/296706@main
